### PR TITLE
Canvas type

### PR DESCRIPTION
**Pull request by lbud**
Closes https://github.com/mapbox/mapbox-gl-js/issues/3580.

This adds a `canvas` source type, which can be used in `raster` layers like existing `image` and `video` sources.

To be resolved:
* We have no solid plan for parity with GL Native for this feature. It's also not something that will be very useful in api-styles, because we're loading same-page-based canvases, unlike other sources which often live in the ☁️. These things are weird and we should make sure we're comfortable with this approach.
* [x] add tests
* [ ] add example

Related feature but not a blocker: raster querying https://github.com/mapbox/mapbox-gl-js/issues/1404 (came up in #3745)

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs (generated)

### CODE REVIEW

**Comment by lucaswoj** [comment]
> Do we _want_ to support `webgl` contexts (less straightforward, not necessarily the use case we're targeting, and likely to be obseleted by custom layer types)?

Based on what I've heard about perf, we don't need to go out of our way to support WebGL canvases. If we get some support for free from `texImage2D`, that's a nice bonus.

**Comment by jliebrand** [comment]
Curious: will this feature allow me to draw to a 2d canvas, and have it show up on the map (correctly zooming/panning/etc)?

Eg will this allow me animate something on the chart, unrelated to geojson and webgl?

**Comment by lucaswoj** [comment]
> Curious: will this feature allow me to draw to a 2d canvas, and have it show up on the map (correctly zooming/panning/etc)?

Yes!

> Eg will this allow me animate something on the chart, unrelated to geojson and webgl?

Yup 😄

**Comment by jliebrand** [comment]
Awesome! Can't wait!

File: package.json [hunk 1]
```diff
@@ -45,6 +45,7 @@
     "babel-eslint": "^7.0.0",
     "benchmark": "~2.1.0",
     "browserify": "^13.0.0",
+    "canvas": "^1.6.2",
```
Comment (lbud) [inline package.json]: NOTE: this is a relatively nontrivial dep for something just used in one test…discuss

Comment (lucaswoj) [inline package.json]: Can you create a fake Canvas API using Sinon?

**Comment by lbud** [comment]
> * As currently implemented, this can support reading from either 2D or webgl canvas contexts. 2D is ideal and more performant and, in fact, gl.texImage2D / gl.texSubImage2D can take an HTML5 canvas element as an argument, meaning we don't even have to manually copy pixels. (When using a webgl context however this causes very incorrect rendering/artifacts.) Do we want to support webgl contexts (less straightforward, not necessarily the use case we're targeting, and likely to be obseleted by custom layer types)?

Alright, I simplified this a lot by using `texImage2D`/`texSubImage2D` no matter the context/not even doing context detection. This also allows us to take out the `dimensions` property since we don't manually `readPixels`. It performs fine on my simple canvas debug pages but I haven't made anything super flashy to stress test it with yet… We _could_ prohibit `webgl` contexts, or we could just _caveat emptor_ it. 🤔 

Once I pull together a good example and tighten up the docs this should be good to 👀 & 🚢 

> * We have no solid plan for parity with GL Native for this feature. It's also not something that will be very useful in api-styles, because we're loading same-page-based canvases, unlike other sources which often live in the ☁️. These things are weird and we should make sure we're comfortable with this approach.

🤔

**Comment by lbud** [comment]
Rebased; I think this is good to go. I actually think this is fine to merge without an example, at least for now. I do have an example I've been playing around with, but it depends on loading 60-some images…this can be a follow-up task 🤔🌩

Secondly, tests were [failing by hanging/timing out at the ignored line-width property function test](https://circleci.com/gh/mapbox/mapbox-gl-js/6603) because the spec `canvas_type` branch is rebased off of current `mb-pages`, which says line-width property functions are not supported (they were rolled back). The suite implementation still renders even ignored tests, and just doesn't count them against the fail count. But in this case, it errors and hangs on validating the line-width property function test. It's not the cleanest, but I [added a commit](https://github.com/mapbox/mapbox-gl-js/pull/3765/commits/51bfa6a021bb043137fab5a8457d497a9dfdf809) to this PR that creates a style separately of the map for each render test and then adds it to the map, so that we can pass in a `validate` boolean based on whether a test is ignored. 👍/:-1:? (This test has been hanging locally for me for weeks 😕)

File: js/source/canvas_source.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            this.height = this.canvas.height;
+            this.resize = true;
+        }
+        if (this._hasInvalidDimensions()) {
+            return this.fire('error', new Error('Canvas dimensions cannot be less than or equal to zero.'));
+        }
+
+        if (!this.tile) return; // not enough data for current position
+        this._prepareImage(this.map.painter.gl, this.canvas, this.resize);
+        this.resize = false;
```
Comment (lucaswoj) [inline js/source/canvas_source.js]: Maybe convert this to a local var?

File: js/source/canvas_source.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            this.width = this.canvas.width;
+            this.resize = true;
+        }
+        if (this.canvas.height !== this.height) {
+            this.height = this.canvas.height;
+            this.resize = true;
+        }
+        if (this._hasInvalidDimensions()) {
+            return this.fire('error', new Error('Canvas dimensions cannot be less than or equal to zero.'));
+        }
```
Comment (lucaswoj) [inline js/source/canvas_source.js]: Maybe make this a no-op instead of a failure

File: test/suite_implementation.js [hunk 4]
```diff
@@ -30,13 +31,15 @@ module.exports = function(style, options, _callback) {
...
-        style: style,
         classes: options.classes,
         interactive: false,
         attributionControl: false,
         preserveDrawingBuffer: true
     });
 
+    const _style = new Style(style, map, { validate: !options.ignored });
+    map.setStyle(_style);
+
```
Comment (lucaswoj) [inline test/suite_implementation.js]: We can use `skipped` flags to get around validating these tests without adding new code

**Comment by lbud** [comment]
> There's a regression test that also uses a data-driven line width. If you skip that one, everything should be 🍏 and you're clear to :shipit:!

aha :female_detective: I saw line-width was still failing and didn't get a chance to look into it yet. Alright, will do!

**Comment by averas** [comment]
This is awesome! One use case I feel I haven't found an optimal solution for has been to effectively show small thumbnail images in Mapbox GL JS, to create maps like this:

screen shot 2015-06-03 at 20 53 08

While it is possible to come up with something quite close to that with the marker plugin I found it hard to make a seamless experience that switches between symbols (dots/icons) and thumbnails based on zoom, and the markers are not data-driven in a similar manner as with other sources which makes it tedious to add and remove them based on map interaction.

Could the <code>canvas</code> source type be used to create a data driven thumbnail map like the one above?

**Comment by lucaswoj** [comment]
@averas 

It would be possible to create a `canvas` source which renders and places all of the thumbnails. I anticipate the difficulty of implementing this (you'll need to work with the canvas API and explicitly do some projections) may outweigh the benefits. 

Have you tried using implementing this using the `image` source?

**Comment by averas** [comment]
@lucaswoj 

> Have you tried using implementing this using the image source?

No..? The <code> image </code> source can only reference a single image, and once created can't be mutated (one has to replace the source to update the image). I imagine it would quite ineffective to push/pop potentially hundreds of sources/layers when panning/zooming a map, but I may be wrong.

I think the best Mapbox GL JS examples of this are:
https://www.mapbox.com/blog/custom-markers-mapboxgl/
https://www.mapbox.com/mapbox-gl-js/example/custom-marker-icons/

.. but again, both of these use makers and are not data driven by sources (i.e. you have to micro manage the markers with discrete code hooked in to map events).

**Comment by yangdonglai** [comment]
marked and keep attention

**Comment by bennlich** [comment]
> (you'll need to work with the canvas API and explicitly do some projections)

@lucaswoj I'm working with a canvas source right now, wondering this exact thing. I how do I project from lat/lng to canvas coordinates?

**Pull request merged**